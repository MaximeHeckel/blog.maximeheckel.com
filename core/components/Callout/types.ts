export type CalloutVariant = 'info' | 'danger';

export interface CalloutProps {
  label?: React.ReactNode | string;
  variant: CalloutVariant;
}
