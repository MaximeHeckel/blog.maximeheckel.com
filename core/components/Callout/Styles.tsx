import { styled } from 'lib/stitches.config';

export const StyledCalloutIconWrapper = styled('div', {
  position: 'absolute',
  display: 'flex',
  top: '-24px',
  right: '-18px',
  borderRadius: '50%',
  padding: '8px',
  color: 'var(--maximeheckel-colors-body)',
  border: '8px solid var(--maximeheckel-colors-body)',
  background: 'var(--icon-background, var(--maximeheckel-colors-body))',

  variants: {
    variant: {
      info: {
        '--icon-background': 'var(--maximeheckel-colors-brand)',
      },
      danger: {
        '--icon-background': 'var(--maximeheckel-colors-danger)',
      },
    },
  },
});

export const StyledCalloutLabelWrapper = styled('div', {
  position: 'absolute',
  display: 'flex',
  top: '-24px',
  right: '-8px',
  borderRadius: 'var(--border-radius-1)',
  padding: '8px',
  color: 'var(--maximeheckel-colors-body)',
  fontSize: 'var(--font-size-1)',
  fontWeight: 'var(--font-weight-3)',
  userSelect: 'none',
  background: 'var(--icon-background, var(--maximeheckel-colors-body))',

  variants: {
    variant: {
      info: {
        '--icon-background': 'var(--maximeheckel-colors-brand)',
      },
      danger: {
        '--icon-background': 'var(--maximeheckel-colors-danger)',
      },
    },
  },
});

export const StyledCallout = styled('aside', {
  '*:last-child': {
    marginBottom: '0px',
  },

  position: 'relative',
  padding: '30px 30px',
  marginBottom: '2.25rem',
  borderRadius: 'var(--border-radius-1)',
  color: 'var(--maximeheckel-colors-typeface-primary)',
  border: '2px solid var(--maximeheckel-colors-foreground)',
  background: 'var(--callout-background, var(--maximeheckel-colors-emphasis))',

  variants: {
    variant: {
      info: {
        '--callout-background': 'var(--maximeheckel-colors-emphasis)',
      },
      danger: {
        '--callout-background': 'var(--maximeheckel-colors-danger-emphasis)',
      },
    },
  },
});
