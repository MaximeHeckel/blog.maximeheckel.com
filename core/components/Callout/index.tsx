export { default } from './Callout';
export type { CalloutProps } from './types';
