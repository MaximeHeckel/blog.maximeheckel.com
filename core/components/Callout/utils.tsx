import { AlertIcon, InfoIcon } from '../Icons';
import { CalloutVariant } from './types';

export const getVariantIcon = (variant: CalloutVariant): React.ReactNode => {
  switch (variant) {
    case 'info':
      return <InfoIcon size={5} />;
    case 'danger':
      return <AlertIcon size={5} />;
  }
};
