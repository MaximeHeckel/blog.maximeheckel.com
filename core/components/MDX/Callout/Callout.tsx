import React from 'react';
import styled from '@emotion/styled';

export enum VARIANT {
  DANGER = 'danger',
  INFO = 'info',
}

interface Props {
  variant: VARIANT;
}

const Callout: React.FC<Props> = (props) => <StyledCallout {...props} />;

const variantColors = {
  [VARIANT.DANGER]: `
    background-color: var(--maximeheckel-colors-danger-emphasis);
    border-color: var(--maximeheckel-colors-danger);
    color: var(--maximeheckel-colors-typeface-primary);
    `,
  [VARIANT.INFO]: `
    background-color: var(--maximeheckel-colors-emphasis);
    border-color: var(--maximeheckel-colors-brand);
    color: var(--maximeheckel-colors-typeface-primary);
  `,
};

const StyledCallout = styled('div')<{ variant: VARIANT }>`
  @media (max-width: 700px) {
    /**
     * Make it fullbleed! 
     */
    width: 100vw;
    position: relative;
    left: 50%;
    right: 50%;
    margin-left: -50vw;
    margin-right: -50vw;

    border-radius: 0px;
  }

  code {
    background: none;
  }

  border-radius: var(--border-radius-0);
  border-left: 3px solid;
  padding: 30px 30px;
  margin-bottom: 2.25rem;

  *:last-child {
    margin-bottom: 0px;
  }

  ${(p) => variantColors[p.variant]}
`;

export { Callout };
