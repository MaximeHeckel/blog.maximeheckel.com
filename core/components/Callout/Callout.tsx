import { Callout as SystemCallout, Flex } from '@maximeheckel/design-system';
import { CalloutProps as SystemCalloutProps } from '@maximeheckel/design-system/dist/types/components/Callout';

const Callout = (props: SystemCalloutProps) => {
  const { children, ...rest } = props;

  return (
    <SystemCallout
      css={{
        marginTop: 'var(--space-3)',
      }}
      {...rest}
    >
      <Flex alignItems="start" direction="column" gap="6">
        {props.children}
      </Flex>
    </SystemCallout>
  );
};

export default Callout;
