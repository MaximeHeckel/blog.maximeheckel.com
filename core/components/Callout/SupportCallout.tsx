import { Anchor, Text } from '@maximeheckel/design-system';
import Callout from './Callout';

const SupportCallout = () => {
  return (
    // @ts-ignore
    <Callout
      label="Support my work"
      css={{
        '--icon-background': 'var(--orange-900) !important',
        '--callout-background': 'var(--orange-400)',
      }}
    >
      <Text>
        <Text as="span" weight="4">
          Enjoying the content and feeling like supporting my work?
        </Text>{' '}
        You can show your appreciation by{' '}
        <Anchor
          href="https://www.buymeacoffee.com/maximeheckel"
          target="_blank"
          rel="noopener noreferrer"
        >
          buying me a coffee
        </Anchor>{' '}
        which will give me the much-needed energy to take on more
        ambitious/high-quality articles and projects.
      </Text>
      <Text weight="4">Thank you for reading!</Text>
    </Callout>
  );
};

export default SupportCallout;
