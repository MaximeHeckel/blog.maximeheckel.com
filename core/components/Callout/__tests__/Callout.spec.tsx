import { cleanup, render } from '@testing-library/react';
import React from 'react';
import Callout from '../Callout';

describe('Callout', () => {
  afterEach(cleanup);
  it('can render a Callout with variant info', () => {
    const { container, baseElement } = render(
      <Callout variant="info">Test</Callout>
    );
    expect(
      container.querySelector('[data-variant="info"]')
    ).toBeInTheDocument();
    expect(container.querySelector('svg')).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it('can render a Callout with variant danger', () => {
    const { container, baseElement } = render(
      <Callout variant="danger">Test</Callout>
    );
    expect(
      container.querySelector('[data-variant="danger"]')
    ).toBeInTheDocument();
    expect(container.querySelector('svg')).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it('can render a Callout with variant info', () => {
    const { container, queryByText, baseElement } = render(
      <Callout label="Test Label" variant="info">
        Test
      </Callout>
    );
    expect(
      container.querySelector('[data-variant="info"]')
    ).toBeInTheDocument();
    expect(container.querySelector('svg')).not.toBeInTheDocument();
    expect(queryByText('Test Label')).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it('can render a Callout with variant danger', () => {
    const { container, queryByText, baseElement } = render(
      <Callout label="Test Label" variant="danger">
        Test
      </Callout>
    );
    expect(
      container.querySelector('[data-variant="danger"]')
    ).toBeInTheDocument();
    expect(container.querySelector('svg')).not.toBeInTheDocument();
    expect(queryByText('Test Label')).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });
});
