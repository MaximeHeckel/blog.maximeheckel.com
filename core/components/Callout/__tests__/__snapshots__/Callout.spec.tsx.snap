// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Callout can render a Callout with variant danger 1`] = `
<body>
  <div>
    <aside
      class="c-dXzGmz c-dXzGmz-jVzdYt-variant-danger"
    >
      <div
        class="c-jBmcOA c-jBmcOA-deNlpD-variant-danger"
      >
        <svg
          class="c-PJLV c-PJLV-bHjGOQ-variant-default c-PJLV-hKgDZV-outline-true c-PJLV-jnsSHc-size-5"
          fill="none"
          height="22"
          role="img"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Alert
          </title>
          <desc>
            An icon representing an exclamation mark in an octogone
          </desc>
          <polygon
            points="7.86 2 16.14 2 22 7.86 22 16.14 16.14 22 7.86 22 2 16.14 2 7.86 7.86 2"
          />
          <line
            x1="12"
            x2="12"
            y1="8"
            y2="12"
          />
          <line
            x1="12"
            x2="12.01"
            y1="16"
            y2="16"
          />
        </svg>
      </div>
      Test
    </aside>
  </div>
</body>
`;

exports[`Callout can render a Callout with variant danger 2`] = `
<body>
  <div>
    <aside
      class="c-dXzGmz c-dXzGmz-jVzdYt-variant-danger"
    >
      <div
        class="c-dxxjkH c-dxxjkH-deNlpD-variant-danger"
      >
        Test Label
      </div>
      Test
    </aside>
  </div>
</body>
`;

exports[`Callout can render a Callout with variant info 1`] = `
<body>
  <div>
    <aside
      class="c-dXzGmz c-dXzGmz-dpIRHz-variant-info"
    >
      <div
        class="c-jBmcOA c-jBmcOA-jXIqcX-variant-info"
      >
        <svg
          class="c-PJLV c-PJLV-bHjGOQ-variant-default c-PJLV-hKgDZV-outline-true c-PJLV-jnsSHc-size-5"
          fill="none"
          height="22"
          role="img"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Info
          </title>
          <desc>
            An icon representing the letter ‘i‘ in a circle
          </desc>
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <line
            x1="12"
            x2="12"
            y1="16"
            y2="12"
          />
          <line
            x1="12"
            x2="12.01"
            y1="8"
            y2="8"
          />
        </svg>
      </div>
      Test
    </aside>
  </div>
</body>
`;

exports[`Callout can render a Callout with variant info 2`] = `
<body>
  <div>
    <aside
      class="c-dXzGmz c-dXzGmz-dpIRHz-variant-info"
    >
      <div
        class="c-dxxjkH c-dxxjkH-jXIqcX-variant-info"
      >
        Test Label
      </div>
      Test
    </aside>
  </div>
</body>
`;
